### PR TITLE
[post] Rector post 2

### PR DESCRIPTION
**Pull request by TomasVotruba**
Ref #221 

Maybe split why (need in industry) + "fouding fathers" and prerequisities (technical part)

### Preview

aaa

### CODE REVIEW

**Comment by TomasVotruba** [comment]
@Lexinek Asking for review in advance :)

**Comment by carusogabriel** [comment]
@TomasVotruba I'll review it this night, no problem?

**Comment by TomasVotruba** [comment]
@carusogabriel No hurry, still got 6 days to go :)

File: source/_posts/2018/2018-02-19-rector-part-2-why.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+related_items: [63, 77] 
+---
+
+## Why was Rector needed - Instant Upgrades? 
+
+2 things you need to upgrade PHP application to new version of the framework you use:
+
+### 1. The Knowledge
+
+Most of you follows changes in Symfony or Nette. You know that [`ProcessBuilder` was removed](https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md#process) and that you'll have to migrate to `Process` class only. You know that [`Nette\Object` was removed](https://forum.nette.org/cs/26250-pojdte-otestovat-nette-2-4-rc) and that you'll have to rewrite all `@method` annotation to real methods.
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-19-rector-part-2-why.md]:
```diff
-Most of you follows
+Most of your follows
```

File: source/_posts/2018/2018-02-19-rector-part-2-why.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+You read blogs, follow news on forum, read all `CHANGELOG.md` and `UPGRADE.md` files and sometimes commits, to find out what all has changed. **You have the knowledge**.
+
+### 2. The Resources
+
+You work in a company where up-to-date it very important value. Your employer finances upgrades and also your education in it (pays you to get the knowledge). Once a 6 month you have dedicated paid time to update all packages to most recent versions. **You have the resources**
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-19-rector-part-2-why.md]:
```diff
-Once a 6 month
+Once a 6 months
```

File: source/_posts/2018/2018-02-19-rector-part-2-why.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+Teams don't have space to find out what changed between Symfony 2.7 and Symfony 2.8. They don't care, because their employer cares mostly about creating that new e-commerce as fast and as cheap as possible. 
+  
+That's why most of lectures (@todo link) I make is about giving teams *the knowledge* that they can use with their very limited resources to make the best out of it.
+ 
+That naturally leads to huge legacy code, team performance drop from 100 % to 20 %, which leads to hiring 5x more people etc. to keep productivity the same, more money, and pressure to faster development, which lead to huge legacy code...
+  
+### Are Deprecations Easy to Find?
+ 
+Let's say you have time to explore the Internet, follow [Symfony News on Twitter](https://twitter.com/symfony_en), read [every news post on Symfony Blog](http://symfony.com/blog/category/living-on-the-edge) or know where on the Nette forum are located [Release Notes](https://forum.nette.org/en/f78-release-announcements-news).
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-19-rector-part-2-why.md]:
```diff
-http://symfony.com
+https://symfony.com
```

File: source/_posts/2018/2018-02-19-rector-part-2-why.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+Sometimes the maintainer goes as far to deprecate code slowly (!= remove):
+
+```php
+trigger_error('Method add() is deprecated, use addHtml() instead.', E_USER_DEPRECATED);
+```
+
+Which is good enough for start. Symfony even has [PHPUnit Bridge](https://symfony.com/doc/current/components/phpunit_bridge.html) that tries to detect those deprecations. Would you know what exactly do you need to change?
+
+<img src="https://symfony.com/doc/current/_images/report.png">
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-19-rector-part-2-why.md]: This `_image` is italicising all the text

File: source/_posts/2018/2018-02-19-rector-part-2-why.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+All this was needed before, but PHP missed crucial parts:  
+
+- parse PHP code to AST - done thanks to Nikic and his [nikic/PHP-Parser](https://github.com/nikic/PHP-Parser) 
+- prototype that allows context aware PHP anylisis = "this variable is of this type" - done in 2017 by Ondřej Mirtes and his [PHPStan](/blog/2017/01/28/why-I-switched-scrutinizer-for-phpstan-and-you-should-too/)
+- save changed AST back to PHP without any changes - [this feature was added in 2017](https://github.com/nikic/PHP-Parser/blob/master/doc/component/Pretty_printing.markdown#formatting-preserving-pretty-printing) to `nikic/php-parser` and becomes stable with version 4 
+- simple coding standard tool, that would cleanup the code after Rector - done in 2017 by [EasyCodingStandard](https://github.com/Symplify/EasyCodingStandard)
+- and prototype which would allow it
+
+### FDD: Friendship-Driven-Development
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-19-rector-part-2-why.md]: This term :heart:

File: source/_posts/2018/2018-02-19-rector-part-2-why.md (hunk 4 above)
Comment (TomasVotruba) [reply source/_posts/2018/2018-02-19-rector-part-2-why.md]: It's just the `_` sign, no problem

**Comment by TomasVotruba** [comment]
@carusogabriel After bit of thinking I split the post into 2. It was too long and questions like why + why now + what needed to happen + what is it solving exactly + why is it needed were mixed together.

Next post is only about pre-requisities and team work

File: source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+That's how [Rector](https://github.com/rectorphp/rector) was born one part after another. At least the ideas.
+
+It's not only about writing code. It's about discussing the idea, finding the right API that not only me but others would understand right away, it's about using other tools. It's about talking with people who already made AST tools, ask for advices and find possible pitfalls. It's about reporting issues and talking with people who maintain  projects you depend on.
+
+### FDD = Friendship-Driven Development
+
+I don't work for at any company so development of Rector doesn't solve my personal issues. That's how most project is born, like PHPStan to check Slevomat's code. That means I needed other motivation - when my frustration of wasted thousands human-hours was not enough. 
+ 
+Here I'd like to thank [Petr Vacha](https://) for cowork weekend in Brno with in summer 2017, where it all started - in those times named as *Refactor*. You've been great friend for years and courage in times, when I needed it the most.
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md]: Missing link

File: source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+You already know [What Rector does and How it works](/blog/2018/02/19/rector-part-1-what-and-how/) from part 1.
+
+It's not that PHP didn't need to be updated until 2017. I surely could delegate hundreds of *upgrade-hours* for my whole career. So **why Now?**
+
+## *Codemods* as Standard
+
+*Codemod* is a tool that modifies your code. And you're ok with it.
+ 
+Many years ago [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) was born by [Greg Sherwood](http://gregsherwood.blogspot.cz/search/label/PHP_CodeSniffer) from Australia. Guess how long ago? In 2006! Tool that checks your coding standard, tabs and spaces, brackets and quotes. First of it's kind to be mainstream in PHP community.
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md]:
```diff
-http://gregsherwood.blogspot.cz
+https://gregsherwood.blogspot.cz
```

File: source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+It was followed by [PHP CS Fixer](https://github.com/friendsofphp/php-cs-fixer) with [it's first release in 2014](http://fabien.potencier.org/php-cs-fixer-finally-reaches-version-1-0.html) by [Fabien Potencier](https://fabien.potencier.org). Did you know the first script [had only 106 lines](https://gist.github.com/fabpot/3f25555dce956accd4dd)?
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md]:
```diff
-https://fabien.potencier.org
+http://fabien.potencier.org
```

File: source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  
+I use daily both of these tools, [they're both awesome and work best together](/blog/2017/05/03/combine-power-of-php-code-sniffer-and-php-cs-fixer-in-3-lines/). Both of them fix the code for you, so you can sleep or have a coffee instead.
+
+It took time and 2 tools with over 4000 stars on Github to get here.
+
+## Is PHP Ready for AST? 
+
+Few year ago [Nikita Popov](https://nikic.github.io/) started ambitious project [nikic/PHP-Parser](https://github.com/nikic/PHP-Parser). PHP-Parser parses PHP code to AST. If you're new to Abstract Syntax Tree (AST), check [this post that describes 2 big changes in PHP ecosystem](/blog/2017/11/06/wow-to-change-php-code-with-abstract-syntax-tree) thanks to AST.
```
Comment (carusogabriel) [inline source/_posts/2018/2018-02-26-rector-part-2-maturity-of-php-ecocystem-and-founding-fathers.md]: "A few _years_ ago started _an_ ambitious project"

**Comment by TomasVotruba** [comment]
Thanks :+1:

**Pull request merged**